src/process/unix/SDL_process.c

```diff
@@ -0,0 +1,314 @@
+        }
+
+        newptr++;
+        ptr++;
+    }
+
+    *newptr = NULL;
+    return newlist;
+}
+
+void freestrlist(char **list)
+{
+  char **ptr = list;
+
+  while (*ptr) {
+      SDL_free(*ptr);
+      ptr++;
+  }
+
+  SDL_free(list);
+}
+
+SDL_Process *SDL_CreateProcess(const char * const *args, const char * const *env, SDL_ProcessFlags flags)
+{
+    // Keep the malloc() before exec() so that an OOM won't run a process at all
+    SDL_Process *process = SDL_malloc(sizeof(SDL_Process));
+
+    if (!process) {
+        return NULL;
+    }
+
+    process->flags = flags;
+
+    if ((flags & SDL_PROCESS_STDIN) && (pipe(process->stdin_pipe) < 0)) {
+        SDL_SetError("Could not pipe(): %s", strerror(errno));
+        SDL_free(process);
+        return NULL;
+    }
+
+    if ((flags & SDL_PROCESS_STDOUT) && (pipe(process->stdout_pipe) < 0)) {
+        SDL_SetError("Could not pipe(): %s", strerror(errno));
+
+        if (flags & SDL_PROCESS_STDIN) {
+            close(process->stdin_pipe[READ_END]);
+            close(process->stdin_pipe[WRITE_END]);
+        }
+
+        SDL_free(process);
+        return NULL;
+    }
+
+    if ((flags & SDL_PROCESS_STDERR) && (pipe(process->stderr_pipe) < 0)) {
+        SDL_SetError("Could not pipe(): %s", strerror(errno));
+
+        if (flags & SDL_PROCESS_STDIN) {
+            close(process->stdin_pipe[READ_END]);
+            close(process->stdin_pipe[WRITE_END]);
+        }
+
+        if (flags & SDL_PROCESS_STDOUT) {
+            close(process->stdout_pipe[READ_END]);
+            close(process->stdout_pipe[WRITE_END]);
+        }
+
+        SDL_free(process);
+        return NULL;
+    }
+
+    process->pid = fork();
+
+    if (process->pid < 0) {
+        SDL_SetError("Could not fork(): %s", strerror(errno));
+        if (flags & SDL_PROCESS_STDIN) {
+            close(process->stdin_pipe[READ_END]);
+            close(process->stdin_pipe[WRITE_END]);
+        }
+
+        if (flags & SDL_PROCESS_STDOUT) {
+            close(process->stdout_pipe[READ_END]);
+            close(process->stdout_pipe[WRITE_END]);
+        }
+
+        if (flags & SDL_PROCESS_STDERR) {
+            close(process->stderr_pipe[READ_END]);
+            close(process->stderr_pipe[WRITE_END]);
+        }
+
+        SDL_free(process);
+
+        return NULL;
+    } else if (process->pid == 0) {
+        if (flags & SDL_PROCESS_STDIN) {
+            close(process->stdin_pipe[WRITE_END]);
+            dup2(process->stdin_pipe[READ_END], STDIN_FILENO);
+        }
+
+        if (flags & SDL_PROCESS_STDOUT) {
+            close(process->stdout_pipe[READ_END]);
+            dup2(process->stdout_pipe[WRITE_END], STDOUT_FILENO);
+        }
+
+        if (flags & SDL_PROCESS_STDERR) {
+            close(process->stderr_pipe[READ_END]);
+            dup2(process->stderr_pipe[WRITE_END], STDERR_FILENO);
+        }
+
+        char **mutable_args = dupstrlist(args);
+        char **mutable_env = env ? dupstrlist(env) : NULL;
```

Comment:
Unfortunately, calling non-async-signal-safe functions like `malloc()` after `fork()` in a multi-threaded process is undefined behaviour. (You'll usually get away with it on Linux with glibc, but not with other libc implementations.)

For the arguments to `execve()` and friends, I *think* it's OK to just cast them - `execve(args[0], (char * const *) args, (char * const *) envp)` - which would avoid needing to dup and free them.

Reply:
My man page decided to hide that quite important information as a single bullet point in the middle of the third list :)

I checked for more info about exec*() and indeed: https://stackoverflow.com/questions/190184/execv-and-const-ness (leaving this here for future reference)

I'll just cast them. If there is any need to play with memory, it can be done before fork() and free'd afterwards.